Add FAB Speed Dial and FAB Toolbar components.

topherfangio opened the pull request:
Fixes #1416 - Adds the Floating Action Button Speed Dial and Toolbar components.

CLA has been signed under Profoundry LLC.

Please let me know if you need any changes! :-)

topherfangio commented:
I signed it!

Signed under company Profoundry LLC with email topher@profoundry.us.

On Wed, May 13, 2015 at 4:31 PM googlebot <notifications@github.com> wrote:

> Thanks for your pull request. It looks like this may be your first
> contribution to a Google open source project, in which case you'll need to
> sign a Contributor License Agreement (CLA).
>
> [image: :memo:] *Please visit https://cla.developers.google.com/
> <https://cla.developers.google.com/> to sign.*
>
> Once you've signed, please reply here (e.g. I signed it!) and we'll
> verify. Thanks.
> ------------------------------
>
>    - If you've already signed a CLA, it's possible we don't have your
>    GitHub username or you're using a different email address. Check your
>    existing CLA data <https://cla.developers.google.com/clas> and verify
>    that your email is set on your git commits
>    <https://help.github.com/articles/setting-your-email-in-git/>.
>    - If you signed the CLA as a corporation, please let us know the
>    company's name.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/material/pull/2842#issuecomment-101822074>.
>

ThomasBurleson commented:
@topherfangio - this looks very impressive.

topherfangio commented:
@ThomasBurleson Thanks for the kind words!

Have you found any issues that need to be addressed?

If there aren't any changes, and I have time before the 0.10.0 release, I may try to add a more generic method that supports both the toolbar and any other "generic material" as shown in the video demonstrating a pop-up menu.

marcysutton commented:
This looks great, just a few comments! 

On the FAB Speed Dial, can we limit it to one direction at a time and slow down the animation? The current animation sequence with all 4 speed dials at once is a bit overwhelming visually, and it also complicates labeling from a screen reader perspective (plus they probably shouldn't be used that way). 

For labeling the speed dial, the triggering FAB needs some kind of toggle language to indicate what it does and whether it's open or closed. `aria-haspopup`, `aria-expanded` and `aria-controls` are a few attributes to consider. If there were more than one on a page, just for context, each instance of the menu would need unique button labels in it to differentiate one from the other.

For the FAB toolbar, obviously we need to fix the icon button colors. I'd look at a similar pattern as I described for the speed dial: associating the triggering button with its associated content with ARIA, and semantically indicating whether it's open or closed. I am totally available to help you work through this, I just wanted to capture some ideas here.

For keyboard focus of both demos, should the FAB behave more like a toggle where the child items are removed from the tab order until explicitly toggled? It's a bad idea to force keyboard users through items that are otherwise accessed through hover. I don't think this level of accessibility detail was reached for the Inbox version (but it needs it).

topherfangio commented:
Hi @marcysutton, thanks for the feedback!

I can definitely update the speed dial demo to only allow one direction at a time, I think this definitely makes a lot more sense. I should be able to update this tonight or tomorrow.

**Re: Labeling** - I can certainly add the `aria` controls, but I could definitely use a bit of direction in this area as I made the assumption it would be up to the developer who uses the component to add the proper aria attributes. Do I just need to update the demo to show how this could be done, or were you suggesting that the speed dial/toolbar actually update the child elements automatically?

**Re: Icon Colors** - I just realized that this didn't follow the same behavior as the standard toolbar. Do I just need to add the `md-focused` class to the currently focused button to fix this?

Lastly regarding keyboard focus, I'm not quite sure I understand the scenario. If we assume a complete lack of a mouse, then the only way to toggle/activate the child elements would be to tab to the FAB at which point it would activate and the children would then be added to the tab index, so you would still have to tab through the children (i.e. there would still be no way to skip them). I admittedly have not done much with keyboard support, so perhaps there is a scenario I am missing regarding screen readers?

Thanks again! Looking forward to making this as awesome as possible :-)

topherfangio commented:
@marcysutton Quick update: I just realized that in Gmail you can't even tab to the compose FAB. Is that the intended behavior? Since it is a button and you can click it, I just assumed you could tab to it, but if that is not the case, we can definitely remove it and it's children from the tabindex.

topherfangio commented:
@marcysutton Sorry for the slew of updates; just noticed what you were talking about regarding the toolbar icon color. I think that broke when I merged in the latest master changes. I'll obviously fix that :-)

ThomasBurleson commented:
*  The toolbar slideout is too fast and [perhaps] should be below the button as the trigger button fades; e.g. an improved transition would be good.
*  To apply a change and trigger a digest/binding updates, use:
```js
vm.focus = function() { $scope.$apply('isOpen = true'); };
```
*  Rename to `md-is-open` since it is a value/state and not an expression
*  The lookup of `backgroundElement` and `triggerElement` are too dependent upon DOM order and will easily fail. These lookups should not be in `$watch()`.

topherfangio commented:
@ThomasBurleson First three points sound good; I'll make the changes.

Regarding DOM lookup, I totally agree but can you suggest a better way? Without jqLite supporting CSS selectors, the only thing I can think of is to create full elements so I can do `$element.find('md-fab-toolbar-background')`. That seemed like a bit of overkill to create a whole element just so I could find it more easily.

ThomasBurleson commented:
How about something like $element.querySelector('div.md-fab-toolbar-background')` as an example.
And if you do this 1x in the postLink, the perf impact is minimal.

ThomasBurleson commented on src/components/fabToolbar/fabToolbar.js:
```diff
@@ -0,0 +1,134 @@
+        isOpen: '=?mdOpen',
+      },
+
+      controller: controller,
```
Rename to `FabToolbarController`

ThomasBurleson commented on src/components/fabToolbar/fabToolbar.js:
```diff
@@ -0,0 +1,134 @@
+    };
+
+    function controller($scope, $element) {
+      var vm = this;
```
No need for this approach since you are not using **controllerAs** syntax now using such notations in your template.

topherfangio replied:
Yeah, I was originally using `controllerAs` but switched when I was having trouble with the scope and wound up using `$timeout`. I'll remove this.

ThomasBurleson commented on src/components/fabToolbar/fabToolbar.js:
```diff
@@ -0,0 +1,134 @@
+
+      vm.focus = function() {
+        // TODO: Why does this need to be inside a $timeout???
+        $timeout(function() {
```
Use `$scope.$apply('isOpen = true')` to update the scope variable and trigger databinding updates.

ThomasBurleson commented on src/components/fabToolbar/fabToolbar.js:
```diff
@@ -0,0 +1,134 @@
+          var scaleHeight = width * 2;
+
+          // Set the positioning and border
+          var borderRadius = width;
```
Not needed.

topherfangio replied:
Can you explain why the border radius is not needed? In the animation video on the design spec, it clearly shows (okay, semi-clearly shows; I had to play/pause the animation about 20 times to see it) a rounded edge as the toolbar appears.

ThomasBurleson replied:
Since you only use it in 
```js 
backgroundElement.style.borderRadius = borderRadius + 'px';
```
I was suggesting just:
```js 
backgroundElement.style.borderRadius = width + 'px';
```

topherfangio replied:
Ah gotcha, will do.

ThomasBurleson commented on src/components/fabSpeedDial/fabSpeedDial.scss:
```diff
@@ -0,0 +1,167 @@
+        &:nth-child(4) { transform: translateX(-$trigger-height * 4 - $trigger-offset - 5); z-index: 3; }
+        &:nth-child(5) { transform: translateX(-$trigger-height * 5 - $trigger-offset - 5); z-index: 2; }
+        &:nth-child(6) { transform: translateX(-$trigger-height * 6 - $trigger-offset - 5); z-index: 1; }
+      }
```
So this styling effectively limits the animations to max of 6 action items ?

topherfangio replied:
Indeed; I tried to refactor this using an SCSS for-loop so I could easily support 100 or more, but it wouldn't recognize a variable inside of the `:nth-child()` pseudo-selector. If there is a better way, I am all for fixing this as it seems stupid to limit it to 6 even if that is what the spec suggests.

ThomasBurleson replied:
Maybe we just predefine for  <= 10 items and then note it in the docs.
@robertmesserle - any thoughts here?

jelbourn replied:
My understanding is that these menus should be *extremely* limited, and 6 actually sounds about right. 

This is actually quite a lot of CSS, though, so I wonder if it would make sense to set the styles w/ javascript.

robertmesserle replied:
I would use an scss loop for this, using the syntax @jelbourn used above for the `nth-child` syntax.

ThomasBurleson replied:
Agreed... but just prebuild the CSS output for 10 items. Right ?

jelbourn replied:
I would lean towards having no CSS for this and instead applying the styles via JS.

topherfangio replied:
Using JS would also allow us to calculate the trigger height rather than specifying it in the SCSS. Could mean supporting more/different/custom FABs. I'll see what I can drum up.

ThomasBurleson commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+
+      compile: function($element) {
+        // Wrap every child in a new div and add a class that we can scale independently
+        var children = $element.find('md-fab-actions').children(),
```
Perhaps better to create a new `md-fab-actions` directive which registers its children elements with the **FabSpeedDialController**. This gives us more control to manipulate and sanity check action items.

topherfangio replied:
Sounds good; I'll get that implemented.

ThomasBurleson commented:
@topherfangio - Really excellent first draft. After you make these changes, I will ask @jelbourn to provide a final review. 

> With these changes I think we will have something solid.

jelbourn commented on src/components/fabSpeedDial/demoBasicUsage/index.html:
```diff
@@ -0,0 +1,69 @@
+<div ng-controller="AppCtrl" layout="column">
+    <md-content class="md-padding" layout="column">
+        <p>
+            You may supply a direction of <code>left</code>, <code>up</code>, <code>down</code>, or <code>right</code>
```
Can you keep your lines to 100 characters (here and everywhere else)?

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+      vm.element = $element;
+
+      // Set the default direction to 'down' if none is specified
+      if (typeof $scope.direction === 'undefined') {
```
Why not just use a falsy check?

topherfangio replied:
Here, you are correct, it is not necessary. Below, however, the isOpen value could legitimately be set to false and we don't want to override it.

jelbourn replied:
For that case you can just use
```js
this.isOpen = this.isOpen || false;
```

You shouldn't need that `$apply`

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+   * trigger element.
+   * @param {expression=} md-open Programmatically control whether or not the speed-dial is visible.
+   */
+  function MdFabSpeedDialDirective($timeout) {
```
Can you refactor the code such that the "directive" is just a simple configuration object, where the `compile`, `link`, and `controller` properties are set as references to names functions?

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+      scope: {
+        direction: '@?mdDirection',
+        isOpen: '=?mdOpen'
+      },
```
Use `bindToController: true` to bind to the controller instance instead of the `$scope` instance.

topherfangio replied:
Just to confirm, you are saying that I should alter the directive/controller/template to use something like `vm.myScopeParam` instead of directly accessing `$scope`? I actually had this first and would be happy to revert to that.

jelbourn replied:
Yes. Using the controller instance instead of the scope instance is much more in-line with how things iwll be in Angular 2 and is the recommendation of the angular team.

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+      // Set the default direction to 'down' if none is specified
+      if (typeof $scope.direction === 'undefined') {
+        $timeout(function() {
+          $scope.direction = 'down';
```
Can you make an enum for the direction values?

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+    };
+
+    function controller($scope, $element) {
+      var vm = this;
```
Can you use `self` instead of `vm` ?

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+        $timeout(function() {
+          vm.scope.isOpen = false;
+        });
+      };
```
`focus` and `blur` here would be much more accurately named `open` and `close`

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+
+    }
+
+    function link($scope, $element, $attrs, ctrl) {
```
These arguments shouldn't start with a `$` because they are not injected. The `link` function is simply passed them.

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    function link($scope, $element, $attrs, ctrl) {
+      var i, trigger, child, children, focusable;
```
Move as much of this code from the `link` function to the controller as possible.

jelbourn replied:
Just declare the variables where they are used. The hoisting behavior only rarely ever affects anything.

robertmesserle replied:
@jelbourn I would disagree with this - it might be personal preference, but I much prefer having a set list up top so that I know the variables being used.

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+      }
+    };
+
+    function controller($scope, $element) {
```
The name of this function should be `MdFabSpeedDialCtrl` or `MdFabSpeedDialController`

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,138 @@
+      trigger.on('blur', ctrl.blur);
+
+      // Add some focus/blur events to whatever (post-compiled) elements the user added in the speed dial
+      children = $element.find('md-fab-actions').children();
```
This seems like it could be its own directive.

topherfangio replied:
@ThomasBurleson said the same thing; I will give this a go when I have a slightly larger chunk of time to dedicate to it :-)

topherfangio replied:
@jelbourn or @ThomasBurleson - I am working on refactoring this into it's own directive. Should it also be in it's own module even though it is a "sub"-directive to the speed dial and toolbar? If so, does it need all of the proper documentation that usually comes with a directive or should I just point to the sd/toolbar docs?

jelbourn replied:
It should go in the same module. The documentation can all live with the top-level API surface.

marcysutton commented:
@topherfangio for keyboard accessibility of the speed dial, let's have it show on focus/hide on blur of the main FAB but only allow arrow keys to navigate between the child buttons...more like a menu. That way you don't force keyboard users through the whole list of child buttons. A good example is this dropdown menu: http://terrillthompson.com/tests/menus/accessible-mega-menu/test.html

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.scss:
```diff
@@ -0,0 +1,167 @@
+      }
+    }
+
+    &.scale {
```
What is `.scale` ?

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.scss:
```diff
@@ -0,0 +1,167 @@
+  align-items: center;
+
+  &.md-opened {
+    &.fling {
```
All CSS classes must have the `md-` prefix.

ThomasBurleson replied:
:+1:  I missed this one.

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.scss:
```diff
@@ -0,0 +1,167 @@
+          transform: translateY(0px) !important;
+        }
+      }
+    }
```
Can you try to restructure these rules to use less nesting? The nesting makes the rules raise the specificity and is a bit slower.

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.scss:
```diff
@@ -0,0 +1,167 @@
+      md-fab-actions {
+        .md-fab-action-item {
+          visibility: visible;
+          transform: translateY(0px) !important;
```
Any use of `!important` should be accompanies by a comment explaining what's going on and why it's necessary.

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.scss:
```diff
@@ -0,0 +1,167 @@
+  &.fling {
+    &.md-down {
+      .md-fab-action-item {
+        &:nth-child(1) { transform: translateY(-$trigger-height * 1 - $trigger-offset); z-index: 6; }
```
These could be done with a scss loop.

jelbourn replied:
All `z-index` properties should be set from (or based on) something from the z-index master list in `variables.scss`

topherfangio replied:
Ah, thanks for the code snippet. I haven't seen the variable interpolation within SCSS before. I'll update this.

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.spec.js:
```diff
@@ -0,0 +1,95 @@
+    inject(function($compile, $rootScope) {
+      el = $compile(template)($rootScope.$new());
+      $rootScope.$apply();
+    });
```
To make testing easier, you can do something like:

```js
  pageScope = $rootScope.$new();
  element = $compile(template)(pageScope);
  scope = element.scope();
  controller = element.controller('mdFabSpeedDial');
```

This will let you make assertions against the directive's controller instance without having to make all direct DOM checks.

gkalpak replied:
Won't `pageScope` be the same as `scope` in the above snippet ?
Did you by any chance mean `scope = element.isolateScope()` ?
<sub>(I haven't looked at the actual code; I'm just making assumptions here (and they might be totally off).</sub>

jelbourn replied:
Yes, I meant `isolatedScope`

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.spec.js:
```diff
@@ -0,0 +1,95 @@
+
+  beforeEach(module('material.components.fabSpeedDial'));
+
+  function setup(template) {
```
`setup` > `compileAndLink`

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.spec.js:
```diff
@@ -0,0 +1,95 @@
+  it('applies a class for each direction', inject(function() {
+    var fab = setup([
+      '<md-fab-speed-dial md-direction="{{direction}}"></md-fab-speed-dial>'
+    ].join("\n"));
```
What is this `join` thing you're doing? Looks like a no-op.

jelbourn commented on src/components/fabSpeedDial/fabSpeedDial.spec.js:
```diff
@@ -0,0 +1,95 @@
+    ].join('\n'));
+
+    fab.find('button').triggerHandler('focus');
+    $timeout(function() {
```
The inside of a `$timeout` doesn't run in unit tests- `$timeout` is replaced with a mock service. 

Instead, call `$timeout.flush()` and then make your assertion.

jelbourn commented on src/components/fabSpeedDial/demoBasicUsage/index.html:
```diff
@@ -0,0 +1,69 @@
+<div ng-controller="AppCtrl" layout="column">
+    <md-content class="md-padding" layout="column">
```
Use 2 spaces as indent (here and in other html files)

jelbourn commented on src/components/fabToolbar/fabToolbar.js:
```diff
@@ -0,0 +1,134 @@
+        '  <div class="md-fab-toolbar-background"></div>',
+        '  <div ng-transclude></div>',
+        '</div>'
+      ].join('\n'),
```
Instead of using `.join()` for multi-line strings, can you follow the style given by the Google js style guide?
http://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml#Multiline_string_literals

jelbourn commented on src/components/fabToolbar/fabToolbar.js:
```diff
@@ -0,0 +1,134 @@
+
+      controller: controller,
+
+      link: link
```
Same thing about `bindToController` and the distinction between `link` and the controller here.

ThomasBurleson replied:
@jelbourn - can you elaborate. Not sure what you mean here.

jelbourn commented on src/components/fabToolbar/fabToolbar.js:
```diff
@@ -0,0 +1,134 @@
+      $scope.$watch('isOpen', function(isOpen) {
+        // Grab the elements
+        var backgroundElement = $element.children().eq(0).children().eq(0)[0];
+        var triggerElement = $element.children().eq(0).children().eq(1).children().eq(0)[0];
```
These lines are super cryptic and very difficult to understand.

jelbourn commented on src/components/fabToolbar/fabToolbar.js:
```diff
@@ -0,0 +1,134 @@
+        // If we find both elements, use them to position the new backgroundt co
+        if (triggerElement && backgroundElement) {
+          // Get our variables
+          var width = $element[0].offsetWidth;
```
This section needs a high-level explanation of what's going on and should also probably be broken out into its own function.

jelbourn commented on src/components/fabToolbar/fabToolbar.scss:
```diff
@@ -0,0 +1,95 @@
+md-fab-toolbar {
+  display: block;
+
+  $custom-duration: $swift-ease-in-duration / 1.5;
```
Add comment explaining the reasoning behind the custom duration.

ThomasBurleson replied:
And per my earlier notes, the animation is too fast.

jelbourn commented:
@topherfangio let me know when you've gone through the comments and are ready for another pass.

gkalpak commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,130 @@
+   * @restrict E
+   *
+   * @description
+   * The `<md-fab-speed-dial>` directive is used present a series of popup elements (usually `<md-button>`s) for
```
_is used **to** present_

gkalpak commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,130 @@
+
+  angular
+    .module('material.components.fabSpeedDial', [])
+    .directive('mdFabSpeedDial', MdFabSpeedDialDirective);
```
I am not sure what is the convention in the `ngMaterial` codebase (and feeling to lazy to go checkit out right now), but I prefer to have non-instantiable functions start with a lowercase.

topherfangio replied:
I was basing this off of the existing directives I saw such as `MdButtonDirective` and `MdBottomSheetDirective`.

ThomasBurleson replied:
We use **uppercamel-case** for Controller and Directive construction functions

gkalpak replied:
My hopes are on @jelbourn to set you straight :stuck-out-tongue:

jelbourn replied:
The convention *should* be that only _constructors_ are PascalCase, where a constructor is is something that's called with the `new` keyword. 

A naming convention (PascalCase here) is the only way to communicate in JS that something is meant to be used as a constructor intended for use with the `new` keyword. This is the way of the Google js style guide and, as far as I know, is generally the accepted pattern among most of the wider JS community.

gkalpak replied:
<sub>_All hope is lost... :(_</sub>

gkalpak commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,130 @@
+          child = children.eq(i);
+
+          child.wrap('<div class="md-fab-action-item">');
+        }
```
Lines 50 - 58 can be replaced with:

```js
$element.find('md-fab-actions').children().
  wrap('<div class="md-fab-action-item">');
```

gkalpak commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,130 @@
+      var vm = this;
+
+      vm.scope = $scope;
+      vm.element = $element;
```
What is purpose of these 2 lines ? Seem totally redundant.

gkalpak commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,130 @@
+
+      // Set the default direction to 'down' if none is specified
+      if (typeof $scope.direction === 'undefined') {
+        $scope.$apply('direction = "down"');
```
Why use `$apply()` ? Why not just set the property ?

topherfangio replied:
Within the if statement, we should be able to. Within the `vm.focus`/`blur`, it wasn't updating the scope until something else triggered a digest. If there is a better way, I can totally change this, but @jelbourn suggested using the `scope.$apply` technique.

gkalpak replied:
`$apply` is indeed necessary to notify Angular when something is invoked _from outside_.
Without getting into much detail, native/jqLite/jQuery event listener callbacks do need the `$apply`.
(Just make sure you are not calling the same functions from somewhere else while a digest cycle is already in progress.)

But they should be eliminated whenever possible (e.g. inside the ifs), because `$apply` does a great lot of stuff, which isn't necessary at that time.

gkalpak commented on src/components/fabSpeedDial/fabSpeedDial.js:
```diff
@@ -0,0 +1,130 @@
+      $scope.$watch('direction', function(newDir, oldDir) {
+        // Add the appropriate classes so we can target the direction in the CSS
+        $element.removeClass('md-' + oldDir);
+        $element.addClass('md-' + newDir);
```
Maybe a little far-fetched, but using `$animate.add/removeClass()` would allow developers to set up nice transition animations. (Not that anyone is ever likely to do so...)

ajoslin commented:
At first sight, this looks way over-engineered to me...

Here's a version of the speed dial in pure css. A little bit of JS is needed for focus-management, but it's trivial: http://codepen.io/andrewtjoslin/pen/WvGaZv

ajoslin commented:
Now that I read the JS more carefully, I see that it's less than I thought. I just read all the comments and thought all this logic was being done in JS. But I see it's actually done well.

My mistake. *retreats back into the shadows*.

topherfangio commented:
@ThomasBurleson There are still a few pending tasks, but I have updated the code with almost all of the suggestions provided.

I could, however, use a bit more guidance on the toolbar animation changes that you suggested. I tried swapping the z-index of the trigger/background so that the trigger appears on top and then fades out, but it looked odd due to the trigger's drop shadow.

Based on the [animation in the spec](http://www.google.com/design/spec/components/buttons-floating-action-button.html#buttons-floating-action-button-transitions), it seems like the trigger itself expands into the toolbar, and only the icon fades. I'm not quite sure if we can achieve this with the current HTML structure, though, as we would have to modify the trigger element itself (in which case the whole toolbar would be clickable) or make the toolbar/background a child of the trigger (in which case it would get cropped to the size of the trigger).

I'm not really sure where to go from here, so I've left it pretty much the same as before until we can hammer this out a bit.

Also, I've modified the animation speed for both directives to use the defaults in `variables.scss`. I think this works well on the toolbar, but on the speed dial it feels a bit slow; particularly for the scale animation.

Thoughts? And thanks!

topherfangio commented:
Quick note: pending tasks are currently being stored in the commit message so that I can more easily track them.

ajoslin commented:
Tophier: set the scale and border radius of the toolbar so it becomes the
same size as the fab, then use a css transition back to default styles when
the toolbar is active. It's some simple math in JS
On Sun, May 24, 2015 at 21:45 Topher Fangio <notifications@github.com>
wrote:

> Quick note: pending tasks are currently being stored in the commit message
> so that I can more easily track them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/material/pull/2842#issuecomment-105124215>.
>

ajoslin commented:
http://codepen.io/andrewtjoslin/pen/ZGBBOW?editors=011

ThomasBurleson commented:
@topherfangio - want to setup and hangout and review this PR together ?

topherfangio commented:
@ThomasBurleson I think that would be a great idea; I should have taken your previous offer for help. I didn't realize there was quite so much left to do.

At the moment, I am fairly booked with a contract, but I'm sure I can find some time to get together. What works best for you? During the day or at night?

Also, you wouldn't happen to be in the San Francisco office would you? I am in town for a couple of weeks on business and would love to meet people for a beer or something :-)

gauravsaini03 commented:
Any update when we will be having this merged ? Really eager to see floating fab buttons

topherfangio commented:
I sat down with @ThomasBurleson yesterday and made some modifications. He requested that I setup a new PR (https://github.com/angular/material/pull/3108) with the changes and close this one so that we can get it merged.

So, it should be done very soon ;-)
